scripts/Engineering/gui/engineering_diffraction/tabs/focus/model.py

```diff
@@ -36,77 +42,114 @@ def focus_run(self, sample_paths, banks, plot_output, instrument, rb_num, spectr
+                    if success:
+                        workspaces_for_run.append(tof_output_name)
+                        # Save the output to the file system.
+                        self._save_output(instrument, sample_path, name, tof_output_name, rb_num)
+                        self._save_output(instrument, sample_path, name, dspacing_output_name, rb_num)
+                    DeleteWorkspace(sample_ws_clone)
+                    DeleteWorkspace(curves_ws_clone)
+                output_workspaces.append(workspaces_for_run)
                 self._output_sample_logs(instrument, run_no, sample_workspace, rb_num)
+
         # Plot the output
         if plot_output:
             for ws_names in output_workspaces:
                 self._plot_focused_workspaces(ws_names)
 
     @staticmethod
     def _run_focus(input_workspace,
-                   output_workspace,
+                   tof_output_name,
                    vanadium_integration_ws,
                    vanadium_curves_ws,
-                   bank,
-                   full_calib_ws=None,
-                   spectrum_numbers=None):
-        kwargs = {
-            "InputWorkspace": input_workspace,
-            "OutputWorkspace": output_workspace,
-            "VanIntegrationWorkspace": vanadium_integration_ws,
-            "VanCurvesWorkspace": vanadium_curves_ws
-        }
-        if full_calib_ws is not None:
-            kwargs["DetectorPositions"] = full_calib_ws
-
-        if bank is None:
-            kwargs["SpectrumNumbers"] = spectrum_numbers
+                   df_kwarg,
+                   full_calib,
+                   region_calib):
+        if input_workspace.getRun().getProtonCharge() > 0:
+            NormaliseByCurrent(InputWorkspace=input_workspace, OutputWorkspace=input_workspace)
         else:
-            kwargs["Bank"] = bank
-
-        try:
-            return EnggFocus(**kwargs)
-        except RuntimeError as e:
-            logger.error(
-                "Error in focusing, Could not run the EnggFocus algorithm successfully for bank "
-                + str(bank) + ". Error Description: " + str(e))
-            raise RuntimeError()
+            logger.warning(f"Skipping focus of run {input_workspace.name()} because it has invalid proton charge.")
+            return False
+        input_workspace /= vanadium_integration_ws
+        ReplaceSpecialValues(InputWorkspace=input_workspace, OutputWorkspace=input_workspace, NaNValue=0,
+                             InfinityValue=0)
+        ApplyDiffCal(InstrumentWorkspace=input_workspace, CalibrationWorkspace=full_calib)
+        ws_d = ConvertUnits(InputWorkspace=input_workspace, Target='dSpacing')
+        focused_sample = DiffractionFocussing(InputWorkspace=ws_d, **df_kwarg)
+        curves_rebinned = RebinToWorkspace(WorkspaceToRebin=vanadium_curves_ws, WorkspaceToMatch=focused_sample)
+        normalised = Divide(LHSWorkspace=focused_sample, RHSWorkspace=curves_rebinned,
```

Comment:
We can avoid making another workspace here if you overwrite `focused_sample` - you should also be able to make use of overridden operators e.g. `focused_sample /= curves_rebinned` - I think this will still show up in the workspace history as `Divide`?

Reply:
Similarly it should also be OK to overwrite the vanadium curves with the rebinned curves?

Reply:
We need to keep the `AllowDifrerentNumberSpectra` kwarg which I don't think we can replicate with the `/` operator

Reply:
Hmm, I thought the `/=` was clever enough to do the required broadcasting - but I haven't tried it!